// @angular/upgrade/esm5/src/common/src/constants.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
export var $COMPILE = '$compile';
export var $CONTROLLER = '$controller';
export var $DELEGATE = '$delegate';
export var $EXCEPTION_HANDLER = '$exceptionHandler';
export var $HTTP_BACKEND = '$httpBackend';
export var $INJECTOR = '$injector';
export var $INTERVAL = '$interval';
export var $PARSE = '$parse';
export var $PROVIDE = '$provide';
export var $ROOT_SCOPE = '$rootScope';
export var $SCOPE = '$scope';
export var $TEMPLATE_CACHE = '$templateCache';
export var $TEMPLATE_REQUEST = '$templateRequest';
export var $$TESTABILITY = '$$testability';
export var COMPILER_KEY = '$$angularCompiler';
export var DOWNGRADED_MODULE_COUNT_KEY = '$$angularDowngradedModuleCount';
export var GROUP_PROJECTABLE_NODES_KEY = '$$angularGroupProjectableNodes';
export var INJECTOR_KEY = '$$angularInjector';
export var LAZY_MODULE_REF = '$$angularLazyModuleRef';
export var NG_ZONE_KEY = '$$angularNgZone';
export var UPGRADE_APP_TYPE_KEY = '$$angularUpgradeAppType';
export var REQUIRE_INJECTOR = '?^^' + INJECTOR_KEY;
export var REQUIRE_NG_MODEL = '?ngModel';
export var UPGRADE_MODULE_NAME = '$$UpgradeModule';
