// @angular/upgrade/esm5/src/common/src/promise_util.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { isFunction } from './util';
export function isThenable(obj) {
    return !!obj && isFunction(obj.then);
}
/**
 * Synchronous, promise-like object.
 */
var SyncPromise = /** @class */ (function () {
    function SyncPromise() {
        this.resolved = false;
        this.callbacks = [];
    }
    SyncPromise.all = function (valuesOrPromises) {
        var aggrPromise = new SyncPromise();
        var resolvedCount = 0;
        var results = [];
        var resolve = function (idx, value) {
            results[idx] = value;
            if (++resolvedCount === valuesOrPromises.length)
                aggrPromise.resolve(results);
        };
        valuesOrPromises.forEach(function (p, idx) {
            if (isThenable(p)) {
                p.then(function (v) { return resolve(idx, v); });
            }
            else {
                resolve(idx, p);
            }
        });
        return aggrPromise;
    };
    SyncPromise.prototype.resolve = function (value) {
        // Do nothing, if already resolved.
        if (this.resolved)
            return;
        this.value = value;
        this.resolved = true;
        // Run the queued callbacks.
        this.callbacks.forEach(function (callback) { return callback(value); });
        this.callbacks.length = 0;
    };
    SyncPromise.prototype.then = function (callback) {
        if (this.resolved) {
            callback(this.value);
        }
        else {
            this.callbacks.push(callback);
        }
    };
    return SyncPromise;
}());
export { SyncPromise };
